### PR TITLE
Add all documentation for contributing documentation

### PR DESCRIPTION
**Pull request by DragaDoncila**
As discussed at the last docs working group meeting, this PR captures the latest changes to all "docs about contributing docs".

There are five main documents which together should help anyone new to napari.org contribute documentation.
- [Contributor's overview](https://github.com/napari/napari.github.io/blob/153d6f7a19fccf7174ab20f52c3a80d7d1a21674/developers/documentation/index.md)
- [Tutorial template](https://github.com/napari/napari.github.io/blob/153d6f7a19fccf7174ab20f52c3a80d7d1a21674/developers/documentation/tutorial_template.md)
- [How to template](https://github.com/napari/napari.github.io/blob/153d6f7a19fccf7174ab20f52c3a80d7d1a21674/developers/documentation/how_to_template.md)
- [Preparing your notebook tutorial](https://github.com/napari/napari.github.io/blob/153d6f7a19fccf7174ab20f52c3a80d7d1a21674/developers/documentation/prepare_notebook_tutorial.md)
- [Submit "git-free" PR how-to](https://github.com/napari/napari.github.io/blob/153d6f7a19fccf7174ab20f52c3a80d7d1a21674/developers/documentation/how_to_submit_pr_online.md)

I've opened this PR here as opposed to in napari/napari so that we can preview it easily. To preview, checkout this branch and use [Jeremy's guide](https://github.com/napari/napari.github.io#setup-dev-environment) to set up your dev environment.

I've placed this "Documentation" section at the top level for now (I actually don't know how to make a subsection?), purely so that we can check what it's like to browse these documents, and whether they are cohesive.

**Feedback needed**:
- Is it cohesive, and easy to browse?
- Can a new user find what they need regardless of which document they first start in?
- Is any content missing?

You're of course welcome to leave any feedback on individual document content as well (note that some links don't yet exist because I don't know exactly where we'll put this), however note that ultimately **this PR will not be merged** - each document will be contributed in a separate PR to napari/napari once we're happy with the overall structure.

Pinging @GenevieveBuckley because I know you had concerns about how this all would tie together. Also pinging @melissawm who I don't think has seen these for a while.

### CODE REVIEW

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_submit_pr_online.md:15]: I'm not sure this document adds a lot of extra benefit, vs linking to the github docs explaining how to open a PR?

File: developers/documentation/how_to_template.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_template.md]: The style guide should come after more important sections, like "Check that your notebook runs", etc.

**Comment by GenevieveBuckley** [comment]
The `index.md` needs restructuring. Our index page needs to be a one page how to guide, with numbered steps so that if you follow them you go from having nothing, to having a useful docs PR submitted to napari.

Right now this is mixed in and buried by explanation about the diataxis framework, and how we like to organize files in napari. That's nice to know, but it's not a concrete step in the how to guide, so we need to move that to make things less confusing.

I'm imagining this type of structure:
- main page: how to guide for contributing to the napari docs (this doesn't exist just yet, I'd say bits of it are in the second half of `index.md`)
  - subpage: template document(s) for users to download (currently this is `tutorial_template` & `how_to_template`)
  - subpage: how to use jupytext, assuming we can't fit it in the template itself, that would be simpler (currently this is in `prepare_notebook_tutorial`)
  - subpage: how to categorize documentation, and where differnt types live on the repository (currently this is `index.md`)

**Comment by GenevieveBuckley** [comment]
I imagine the main page steps would be roughly these:

## How to write a tutorial or how-to guide for napari

1. Setup: you need to have napari + jupyter installed
    - link to napari user installation docs (clickable link)
2. Download this template (clickable link)
    - Don't know which template you need? Read this guide (clickable link)
3. Open the template notebook with jupyter and add your content
    - Explanation of nbscreenshot and other utility functions (either in the template document, or in a separate document linked to from here)
4. Quality check: does your finished notebook run end to end without errors? ("Restart and Run All" cells). This includes any necessary data downloading, etc.
5. Preparing to upload. Using jupytext to convert your finished notebook into the right format
6. Making a pull request to the napari docs
    - Where to put your new documentation page in the repository
    - Mechanics of how to open a PR (links to github docs, etc)
    - How to see a preview of your documentation page online
7. Iteration based on reviewer feedback (I really like that this was included as a step in some of Draga's work)

**Comment by GenevieveBuckley** [comment]
More minor points I'm happy for other people to disagree with, just throwing these out there:
- Both the `how_to_template` and `tutorial_template` documents seem almost entirely identical. I think only some of the headins are slightly different? Maybe these could be combined? (I know I previously said some duplication was ok, and it is, but this looks like quite a lot of duplication so maybe there is an opportunity there)
- I personally would probably prefer it if most of the "prepare_notebook_tutorial" content was included with the template. I feel like it would make it much more likely for people to see it exists. It's much easier to copy a cell and make a small edit if it's already in the notebook (especially for things like tagging cells, where you'd otherwise have to dig around for the right menu if you're not familiar with making them)

File: developers/documentation/index.md
Comment (GenevieveBuckley) [inline developers/documentation/index.md]: This section needs links that say "Click here to download the template!" (that you know, download the template when you click on them). I see links below, but they look like they'll just open up another browser tab instead of downloading the template.

**Review by GenevieveBuckley — CHANGES_REQUESTED** [review]
See my comments above about the structural changes I think we need to make. Basically, I think the `index.md` file needs the explanation bits pulled out of it, so it can become a straight how-to guide.

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [reply developers/documentation/how_to_submit_pr_online.md:15]: DIscussed at the docs meeting today, Draga says the github docs don't quite have what we want - a guide on how to do this without using git in the terminal.

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_submit_pr_online.md:20]: I think this document would benefit from a short table of contents at the very top, just to get an overview of all the steps. There's a lot of large screenshots in this documents so you have to scroll a long way to see all the headings.

You can make internal hyperlinks in markdown like this:
```
# Title
## Contents
- [Subheading Number 1](#subheading-number-1)
- [Another subheading: Just for fun](#another-subheading-just-for-fun)

## Subheading Number 1
... some text ...

## Another subheading: Just for fun
... text...
```

Note:
- You need to replace spaces with hyphens in the round brackets
- You can't have any capital letters in the text in the round brackets, you have to use all lowercase letters here instead (even if the actual subheading uses capital letters)
- You also can't have an punctuation characters (except the first `#` and hyphens) in the round brackets. So any colons, semicolons, etc. get omitted.

**Comment by DragaDoncila** [comment]
@GenevieveBuckley @melissawm I've made the updates we discussed at the docs wg meeting last week. I'd appreciate if you could have a look and see if we prefer this direction.

One thing I can't figure out is how to make the "Download this template" link an actual file download button? It seems like it should be easy but I've never done it in markdown...

**Review by GenevieveBuckley — COMMENTED** [review]
(Turns out an older comment here is still listed as "pending", so I'm just trying to release it now)

**Comment by GenevieveBuckley** [comment]
> One thing I can't figure out is how to make the "Download this template" link an actual file download button? It seems like it should be easy but I've never done it in markdown...

You can use the markdown link syntax, like this:
```
[Click here to download this template](https://link-to-download-file)
```

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_submit_pr_online.md:47]: This section needs to mention that we're asking people to upload a markdown file (so, not a `.ipynb` file). 

You may also need to link to the part of the docs that explains how to convert file formats using jupytext.

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_submit_pr_online.md:106]: > This automatically updates your pull request with the latest changes to the file.

This is a great point to include!

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_submit_pr_online.md]:
```suggestion
Once your pull request is approved, a napari core developer will merge it, and your document will be visible on [napari.org](https://napari.org/)!
```

File: developers/documentation/index.md
Comment (GenevieveBuckley) [inline developers/documentation/index.md]: I agree that we should have this section here, but I don't want "TODO" or "coming soon" in the documentation. I'd prefer we leave out those references until it's time, and make an issue to track the undone task if needed. In this case, we could add an extra task to the existing issue for previewing documentation.

File: developers/documentation/how_to_submit_pr_online.md
Comment (GenevieveBuckley) [inline developers/documentation/how_to_submit_pr_online.md:31]: Honestly, I suspect the best first step is to ask people to open an issue to suggest the new documentation/tutorial they're thinking of adding, to see if there's enthusiasm before putting in a lot of work.
Do you have thoughts, @sofroniewn & @jni?

File: developers/documentation/docs_template.md
Comment (GenevieveBuckley) [inline developers/documentation/docs_template.md:128]: Should we have a short list of common functions that might get used in hidden cells? I'm thinking very small, just the list & links to those functions in the documentation, and only for the types of things people are likely to adjust without thinking about in the GUI and might not know how to do programmatically.

These might include:
- resizing a window (like you have already mentioned here)
- Toggling the layer visibility (eg: several layers are displayed, but we want only one to show up in the screenshot)
- Setting the dims slider position
- Adjusting the contrast
- Switching the viewer mode (2D -> 3D, or gridview)
- ...anything else I'm forgetting?

**Comment by GenevieveBuckley** [comment]
Thanks @DragaDoncila! This is looking much more streamlined and easier to use.

File: developers/documentation/how_to_submit_pr_online.md
Comment (sofroniewn) [reply developers/documentation/how_to_submit_pr_online.md:31]: Yeah we could have something like `0. Before you start`. I don't think making an issue is required, but I do think it should be encouraged, especially if more than a small contribution. Typos, clear fixes etc are fine without issues. Adding something to that effect might help

File: developers/documentation/how_to_submit_pr_online.md
Comment (DragaDoncila) [reply developers/documentation/how_to_submit_pr_online.md:20]: Oops sorry I missed that one @GenevieveBuckley. What level of subheading were you hoping to go down to in the table? The new theme has `##` headings on the right navigation which scrolls with you e.g.:

image

Do you think this would be sufficient?

File: developers/documentation/docs_template.md
Comment (DragaDoncila) [reply developers/documentation/docs_template.md:128]: I've added a list of these settings. Where I could find the property/function in our API reference I've linked it. Where I couldn't, I've provided a code snippet to show how the setting could be used.

File: developers/documentation/how_to_submit_pr_online.md
Comment (DragaDoncila) [reply developers/documentation/how_to_submit_pr_online.md:31]: Thanks for the feedback @GenevieveBuckley and @sofroniewn - I've added a 0th step guiding users to open an issue if they'd like to add a new document

**Comment by DragaDoncila** [comment]
@GenevieveBuckley I think this is now ready for a final pass before we close and reopen in `napari/napari` to actually contribute it. However, it would be nice to get a review from someone outside the docs working group - I don't think we've really shared this work with the community. 

@sofroniewn @jni I'd love if you could look through this PR - these documents would be all the content available for users wanting to contribute documentation. It's in this PR so it can be viewed as a whole, and so that it can be easily previewed with the napari.org theme.

- Do you have any big concerns with the flow of the content?
- Is anything glaringly missing?

File: developers/documentation/index.md
Comment (sofroniewn) [inline developers/documentation/index.md:118]: It might be great to provide some additional info either here - or through links, elsewhere - that lets folks know how to pick between these three.

I could even imagine mentioning this up at the top, say in "0. before you start". I'll note above

File: developers/documentation/index.md
Comment (sofroniewn) [inline developers/documentation/index.md:29]:
```suggestion
Our main types of documents are the following:
- **Explanations** that live in `napari/docs/guides/`
- **Tutorials** that live in `napari/docs/tutorials/`
- **How-tos** that live in `napari/docs/howtos/`

If you're unsure about where your document should live ....
```

**Review by sofroniewn — APPROVED** [review]
Looks great! I love the detail. I flagged one fairly minor point in a comment.

For the main PR are we going to use the trick of uploading the images in the PR comments then linking to the images through the web (rather than add them directly to the repo). This will keep size down in the main repo.

I'm also wondering if we should add anything about `.png` files that people might want to add (we discourage it for reasons about size mentioned above). Also maybe some comments that we don't want data added to the main repo. Maybe that's obvious, but it doesn't hurt to comment as such.

File: developers/documentation/index.md
Comment (DragaDoncila) [reply developers/documentation/index.md:118]: Yes we initially had more content around that but found it difficult to settle on guidelines before we finish our existing content re-org.

I agree with you we needed *something* though. I've added brief sentences about each of the different types of documents in the "Before you start" section. Thanks for catching that.

**Comment by DragaDoncila** [comment]
> For the main PR are we going to use the trick of uploading the images in the PR comments then linking to the images through the web (rather than add them directly to the repo). This will keep size down in the main repo.

@sofroniewn yes and thank you for reminding me! I've also added text to that effect to the instructions for users

> Also maybe some comments that we don't want data added to the main repo. Maybe that's obvious, but it doesn't hurt to comment as such.

Agreed. I've flagged this now in the pre-requisites list of the docs template.

**Comment by GenevieveBuckley** [comment]
I notice this PR is still marked as a draft, @DragaDoncila are you waiting on a few things here, or is this mostly ready (it seems mostly ready to me, my understanding is that all the review comments have been addressed & we could probably merge this whenever you'd like)

**Comment by DragaDoncila** [comment]
@GenevieveBuckley this is still in draft mode because I'm waiting on some of the dust to settle from the TOC reorg before moving this PR into `napari/napari` where it actually belongs - I only opened it here to give us quick access to previews.

File: developers/documentation/docs_template.md
Comment (neuromusic) [inline developers/documentation/docs_template.md:35]: based on https://github.com/napari/napari/pull/3815, we should also make it clear _where_ such images should be stored (probably not in the repo, especially if we move forward with `napari/napari` as the location of the docs)

Comment (sofroniewn) [reply developers/documentation/docs_template.md:35]: yeah, we might want to put this in extra-highlights, and even add to our core developer guide to look out for this in review as the negative consequences of this can be quite large (having to remove commit from GitHub history). Thanks for flagging @neuromusic!!

**Comment by GenevieveBuckley** [comment]
One extra point, is that images should have meaningful alt-text. This should be communicated to users, and also included as a reviewer checklist item.

We're working on a mechanism for alt-text with `nbscreenshot` here: https://github.com/napari/napari/pull/3825#issuecomment-996040101
This is all a little bit in flux at the moment, so it's fine to make a follow up PR if needed. Just flagging it for you now.

**Comment by psobolewskiPhD** [comment]
@DragaDoncila Some good things here, like the `PR online` part and some stuff is already more or less there at: https://napari.org/dev/developers/documentation/index.html
Maybe worth trying to port some of this to https://github.com/napari/docs ?
Eitherway, this will need to be closed.

**Comment by psobolewskiPhD** [comment]
As per my comment above, this won't be merged here so closing.
I made an issue over at napari/docs to look into porting stuff from this PR though:
https://github.com/napari/docs/issues/330

**Pull request closed without merging**
